### DISCUSSION TITLE
How do you remove a query parameter?

### DISCUSSION THREAD

**Discussion by nikkwong** [expo/router#661] [Q&A]
I notice that there's `setParams()` in expo-router, but it doesn't have a way to clear current query parameters. I can pass an empty object to setParams like `setParams({})` but the items in the query remain. How can this be achieved?

**Comment by nikkwong** [comment]
Ping

**Reply by marklawlor** [reply]
Set the value to `undefined` 

`setParams({ "remove-this-value": undefined })`

**Reply by nikkwong** [reply]
This, for a brief period, actually triggers a re-render with the param set as the string "undefined" in the router.

```
const { remove_this_value } = useSearchParams()
console.log(remove_this_value); // will be "undefined" for at least one render
```

Which I can hack to get around, but I'm hoping this api improves in the future. Thanks!

**Reply by andredewaard** [reply]
I also get 'undefined' as value for the param, any way to get around this and completely remove the param?

**Comment by trymbill** [comment]
If setting the value to `undefined` is the solution then that should probably be documented and the types updated, right? https://github.com/expo/router/blob/f92f69943a05ef4b2ef66837a75996ceb3ef9f72/packages/expo-router/src/types.ts#L29

**Reply by marklawlor** [reply]
Yes, it appears that was missed. It will be fixed in this PR https://github.com/expo/expo/pull/25654

**Reply by trymbill** [reply]
It doesn't look like it's just the types that are off, I've tried setting a query parameter to `null` and `undefined` and it just sets it to the stringified version of `"null"` and `"undefined"` respectively 🤔

**Reply by NilsBaumgartner1994** [reply]
Might be related to this issue: https://github.com/expo/expo/issues/26565 ?

**Comment by codinsonn** [comment]
Am I correct in assuming you cannot yet clear params in SDK 51 by either
- passing an empty object?
- setting the value to `null` or `undefined`?

Tried today, but to no avail. Would love to hear of any other solutions that worked for you.

**Reply by NilsBaumgartner1994** [reply]
’’’
router.setParams({test: undefined})
’’’

That's how it works for me

**Reply by NilsBaumgartner1994** [reply]
But yes, I was also facing some weird bugs with Params.

**Comment by komfysach** [comment]
Also facing the same issue, is there an update on this?

**Comment by erickreutz** [comment]
This still seems to be an issue? `router.setParams({ search: undefined })` results in search to be `"undefined"`

**Comment by twoodruff01** [comment]
Just stumbled across this with the exact same problem as described by @erickreutz

**Comment by mysa-mudit** [comment]
still looking for a solution for this

**Comment by billsbooth** [comment]
Yep, still doesn't work 😭

I am writing a deep link handler, and wanted to clear the param when I have handled the link; since removing the param doesn't actually remove it, it ends up in an infinite loop!

**Comment by dineshsalunke** [comment]
For anyone still struggling with this issue here is the patch which stops the undefined and null from converting to string

```
diff --git a/build/hooks.js b/build/hooks.js
index 55f5792af2d1b81cfa92e7ee7e57744bba4c0b71..5695e4933a8a2fba98e71fbf0e23ced129d7c4bf 100644
--- a/build/hooks.js
+++ b/build/hooks.js
@@ -134,7 +134,9 @@ function useLocalSearchParams() {
                     }
                 }),
             ];
-        }
+       } else if ( typeof value === 'undefined' || value === null) {
+           return [key, value]
+       }
         else {
             try {
                 return [key, decodeURIComponent(value)];

```

**Reply by dineshsalunke** [reply]
I have raised a pull request with the above patch

https://github.com/expo/expo/pull/34950
